lib/wallaroo/core/topology/router.pony

```diff
@@ -844,49 +1042,48 @@ class val DataRouter is Equatable[DataRouter]
     ifdef "trace" then
       @printf[I32]("Rcvd msg at DataRouter\n".cstring())
     end
-    let target_id = d_msg.target_id()
     try
-      let target = _data_routes(target_id)?
       ifdef "trace" then
         @printf[I32]("DataRouter found Step\n".cstring())
       end
-      try
-        let route_id = _target_ids_to_route_ids(target_id)?
-        d_msg.deliver(pipeline_time_spent, target, producer_id, producer,
-          seq_id, route_id, latest_ts, metrics_id, worker_ingress_ts)
-        ifdef "resilience" then
-          producer.bookkeeping(route_id, seq_id)
-        end
-      else
-        // This shouldn't happen. If we have a route, we should have a route
-        // id.
-        Fail()
+      let route_id = d_msg.deliver(pipeline_time_spent, producer_id, producer,
+        seq_id, latest_ts, metrics_id, worker_ingress_ts,
+        _data_routes,
+        _target_ids_to_route_ids,
+        _route_ids_to_target_ids,
+        _keyed_routes,
+        _keys_to_route_ids
+         )?
+      ifdef "resilience" then
+        producer.bookkeeping(route_id, seq_id)
       end
     else
-      Fail()
+      // !@ We used to fail here, but the only reason we should get here is
+      // is because we couldn't find a route for the key. Maybe we should
+      // handle this differently.
+      ifdef "trace" then
+        @printf[I32]("DataRouter could not find a step\n".cstring())
+      end
     end
 
   fun replay_route(r_msg: ReplayableDeliveryMsg, pipeline_time_spent: U64,
     producer_id: StepId, producer: DataReceiver ref, seq_id: SeqId,
     latest_ts: U64, metrics_id: U16, worker_ingress_ts: U64)
   =>
     try
-      let target_id = r_msg.target_id()
-      let route_id = _target_ids_to_route_ids(target_id)?
-      //TODO: create and deliver envelope
-      r_msg.replay_deliver(pipeline_time_spent, _data_routes(target_id)?,
-        producer_id, producer, seq_id, route_id, latest_ts, metrics_id,
-        worker_ingress_ts)
+      let route_id = r_msg.replay_deliver(pipeline_time_spent,
+        _data_routes, _target_ids_to_route_ids,
+        producer_id, producer, seq_id, latest_ts, metrics_id,
+        worker_ingress_ts, _keyed_routes, _keys_to_route_ids)?
       ifdef "resilience" then
         producer.bookkeeping(route_id, seq_id)
       end
-      false
     else
+      // !@ this now gets an exception on an unknown key, which gets handled
+      // by the StateStepCreator.
       ifdef debug then
         @printf[I32]("DataRouter failed to find route on replay\n".cstring())
       end
```

Comment:
It won't matter because this will change, but we shouldn't remove Fail() in a case like this.

Reply:
This isn't actually a failure anymore, which is why the `Fail()` was removed. Getting an error here means that the there was an unknown key and the `StateStepCreator` is going to fix that. When we handle #2257 this should become clearer in the code, but for now I'll leave a comment noting that this isn't something that needs to be handled here.